File: advent-of-code/c++/src/adventofcode/advent_of_code_solution.cpp
#include "advent_of_code_solution.hpp"
#include "advent_of_code_puzzle.hpp"

#include <format>

template <typename TPart1, typename TPart2>
void AdventOfCodeSolution<TPart1, TPart2>::solve(int day, TPart1 expectedPart1Test, TPart2 expectedPart2Test) {
    this->log(std::format("Day {}", day));
    AdventOfCodePuzzle* adventOfCodePuzzle = new AdventOfCodePuzzle(day);

    this->log("Part 1");
//    this->checkPart([adventOfCodePuzzle] () {
//            return adventOfCodePuzzle->part1TestFile();
//        },
//        [this] (std::vector<std::string> data) {
//            return this->part1(data);
//        }, expectedPart1Test);

    std::vector<std::string>testInput = utils::readInput(adventOfCodePuzzle->part1TestFile());
    auto testPart = this->part1(testInput);
    this->log(std::format("test: {}", testPart));
    this->check(testPart == expectedPart1Test, std::format("solution was {}, expected {}", testPart, expectedPart1Test));

    std::vector<std::string> input = utils::readInput(adventOfCodePuzzle->inputFile());
    this->log("result: ", this->part1(input));

    this->log("Part 2");
    std::vector<std::string>testInputPart2 = utils::readInput(adventOfCodePuzzle->part2TestFile());
    auto testPart2 = this->part2(testInputPart2);
    this->log(std::format("test: {}", testPart2));
    this->check(testPart2 == expectedPart2Test, std::format("solution was {}, expected {}", testPart2, expectedPart2Test));

    this->log("result: ", this->part2(input));
}

template <typename TPart1, typename TPart2>
void AdventOfCodeSolution<TPart1, TPart2>::checkPart(std::function<std::string()> puzzleInput,
                                                     std::function<auto(std::vector<std::string> data)> puzzlePart,
                                                     auto expectedTest) {
    std::string input = utils::readInput(puzzleInput());
    auto part = puzzlePart(input);
    this->log("test: {}", part);
    this->check(part == expectedTest, std::format("solution was {}, expected {}", part, expectedTest));
}

template <typename TPart1, typename TPart2>
void AdventOfCodeSolution<TPart1, TPart2>::check(bool test, std::string message) {
    if(!test) {
        throw std::runtime_error(message);
    }
}

template class AdventOfCodeSolution<int, int>;
template class AdventOfCodeSolution<long, long>;

File: advent-of-code/python/src/adventofcode/advent_of_code_solution.py
from .utils import readInput, log

class AdventOfCodeSolution:
    def __init__(self):
        pass

    def part1(self, data):
        return None

    def part2(self, data):
        return None

    def solve(self, day, expected_part1_test, expected_part2_test):
        advent_of_code_puzzle = AdventOfCodePuzzle(day)
        log(f"Day {day}")

        log("Part 1")
        self.check_part(advent_of_code_puzzle.part1_test_file, self.part1, expected_part1_test)
        input = readInput(advent_of_code_puzzle.input_file())
        log(f"result: {self.part1(input)}")

        if expected_part2_test is not None:
            log()
            log("Part 2")
            self.check_part(advent_of_code_puzzle.part2_test_file, self.part2, expected_part2_test)
            log(f"result: {self.part2(input)}")

    def check_part(self, puzzle_input, puzzle_part, expected_test):
        input = readInput(puzzle_input())
        part = puzzle_part(input)
        log(f"test: {part}")
        self.check(part == expected_test, f"solution was {part}, expected {expected_test}")

    def check(self, test, message):
        if not test:
            log(message)
            raise ValueError(message)


class AdventOfCodePuzzle:
    def __init__(self, day):
        self.__day = day
        self.__input_file = f"Day{day:02}"

    def input_file(self):
        return self.__input_file

    def part1_test_file(self):
        return f"{self.input_file()}_test"

    def part2_test_file(self):
        return self.part1_test_file()
